### PR TITLE
docs: Created quality assurance table for List All Task API 

### LINKED ISSUE #758: Create Quality Assurance table for List All Tasks API
## Description
As a developer,
I need to know what are the test cases for a certain API,
so that I can write new unit tests.

As a tester,
I need to know what are the test cases for a certain API,
so that I can perform manual tests on the API.

The task is to come up and add a quality assurance table for "List All Tasks" API to the docs in the /docs folder.

Refer to #757, if the following task has not been performed. Add all tasks' quality assurance tables under one heading. Currently, the Update task is under heading "Mentorship Relation" and Create Tasks under heading "Tasks". I suggest make the subheading Tasks under Mentorship Relation and add all Tasks' API under it.

#### Note: This issue is labeled for Open Source Hack.
I can assign this issue to you on September 1st as the official opening date for Open Source Hack. Please free to check out other issues in the meantime. Program website: https://anitab-org.github.io/events/open-source-hack/

## Mocks
As done under [docs/quality-assurance-test-cases.md](https://github.com/anitab-org/mentorship-backend/blob/develop/docs/quality-assurance-test-cases.md)

mentorship-backend

### Update [Required]
- [ ] Add test cases for List All Tasks API.
### Enhancement to Update [Optional]
- [ ] Add all tasks' quality assurance tables under one heading.

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 2 mentors.

## Estimation
[4+] hours

### PR DESCRIPTION
**Pull request by khushishikhu**
### Description

Created quality assurance table for List All Task API 

Fixes #758 

### Type of Change:

- Code
- Quality Assurance
- Documentation

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] Update Postman API at /docs folder
- [x] Update Swagger documentation and the exported file at /docs folder

### CODE REVIEW

**Comment by khushishikhu** [comment]
please someone review my PR

File: docs/quality-assurance-test-cases.md [hunk 1]
```diff
@@ -173,6 +173,21 @@ This document contains some examples of test cases for each feature implemented
 | Logged in user tries to complete a task from an non existing request (The request  exists in a different relationship) | Fail |
 | Logged in user tries to complete a task from a request which is not in the ACCEPTED state (as a mentor or as a mentee) | Fail |
 
+### List All Tasks
+**Service:** PUT /mentorship_relation/{request_id}/task/{task_id}/complete
+
+| Test Case      | Outcome |
```
Comment (sumit-badsara) [inline docs/quality-assurance-test-cases.md]: @khushishikhu  These test cases according to me, aren't valid for list all tasks.
This is not similar to what is given as a mock in PR.
@khushishikhu The state of task shouldn't be included in any condition.
Only the owner (mentor/mentee) of a task should be the primary condition to check if the task belongs to the user, if yes, it is returned in the list.
For eg.

Logged in user requests list all tasks, and out of all tasks, those which involve user (as a mentor or as a mentee) - **Success**
and the other one being where user isn't involved - **Fail**

File: docs/quality-assurance-test-cases.md [hunk 2]
```diff
@@ -173,6 +173,21 @@ This document contains some examples of test cases for each feature implemented
 | Logged in user tries to complete a task from an non existing request (The request  exists in a different relationship) | Fail |
 | Logged in user tries to complete a task from a request which is not in the ACCEPTED state (as a mentor or as a mentee) | Fail |
 
+### List All Tasks
+**Service:** PUT /mentorship_relation/{request_id}/task/{task_id}/complete
```
Comment (sumit-badsara) [inline docs/quality-assurance-test-cases.md]: This is not the correct endpoint for list all tasks.

**Review by sumit-badsara — COMMENTED** [review]
You should also try to include the last part (Create a task) under Task heading in Mentorship relation, as mentioned in the PR.

File: docs/quality-assurance-test-cases.md (hunk 1 above)
Comment (khushishikhu) [reply docs/quality-assurance-test-cases.md]: you mean for listing tasks , there should not be a condition of ACCEPTED task or REJECTED task

**Comment by khushishikhu** [comment]
@BadduCoder  i have tried to recover the error that you asked. please review the changes

File: docs/quality-assurance-test-cases.md [hunk 3]
```diff
@@ -173,7 +173,29 @@ This document contains some examples of test cases for each feature implemented
...
+| Test Case  | Outcome |
+| ---------- | -----------|
+| Logged in user requests list all task,or out of all task, those which involves the user (as a mentor or as a mentee) | Success |
+| Logged in user requests list all task,or out of all task, those which involves the user (neither as a mentor nor as a mentee) | Fail |
+| Logged in user requests to list all task whose id does not exist, which involves the user (as a mentor or as a mentee) | Fail |     
+| Logged in user requests list all task,or out of all task, those which doesn't involves the user (as a mentor or as a mentee) | Fail |
+| Logged in user requests to list all task which is invalid (not an integer), which involves the user(as a mentor or as a mentee) | Fail |
+| Not logged in user (invalid token) tries to List all task| Fail |
+
+### Tasks
```
Comment (sumit-badsara) [inline docs/quality-assurance-test-cases.md:177]: This should come even before List all tasks.
`Task` will contain `Create` and `List All Tasks` as subheadings.

File: docs/quality-assurance-test-cases.md [hunk 4]
```diff
@@ -173,7 +173,29 @@ This document contains some examples of test cases for each feature implemented
 | Logged in user tries to complete a task from an non existing request (The request  exists in a different relationship) | Fail |
 | Logged in user tries to complete a task from a request which is not in the ACCEPTED state (as a mentor or as a mentee) | Fail |
 
+### List All Tasks
+**Service:** PUT /mentorship_relation/{request_id}/task/{task_id}
```
Comment (sumit-badsara) [inline docs/quality-assurance-test-cases.md]: This is still the wrong endpoint. 
If you have doubts, you can always refer the docs.
i.e for this case : **GET `/mentorship_relation/{request_id}/tasks`** will be the endpoint.
As you can see on https://mentorship-backend-temp.herokuapp.com/

Check the first endpoint in below image.

Mentorship

File: docs/quality-assurance-test-cases.md [hunk 5]
```diff
@@ -173,7 +173,29 @@ This document contains some examples of test cases for each feature implemented
...
 
+| Test Case  | Outcome |
+| ---------- | -----------|
+| Logged in user requests list all task,or out of all task, those which involves the user (as a mentor or as a mentee) | Success |
+| Logged in user requests list all task,or out of all task, those which involves the user (neither as a mentor nor as a mentee) | Fail |
+| Logged in user requests to list all task whose id does not exist, which involves the user (as a mentor or as a mentee) | Fail |
```
Comment (sumit-badsara) [inline docs/quality-assurance-test-cases.md]: As there will be no `task_id` the test cases will resolve to following cases : 
1. Logged in user requests all tasks for a mentorship relation in which user is involved | Success
2. Logged in user requests all tasks for a mentorship relation in which user isn't involved | Fail
3. Logged in user requests all tasks for a invalid mentorship relation | Fail
4. Not logged in user requests all tasks for any mentorship relation | Fail
5.  ...

These are just rough approximation of the cases, with rough language.  
If the cases aren't clear, you can refer the backend function and list the cases accordingly.

**Review by sumit-badsara — CHANGES_REQUESTED** [review]
I guess this will give you an idea about the further required changes. Best of luck! :fire:

**Comment by khushishikhu** [comment]
> I guess this will give you an idea about the further required changes. Best of luck! 🔥

yeah sure, Thank you for guidance

File: docs/quality-assurance-test-cases.md (hunk 5 above)
Comment (khushishikhu) [reply docs/quality-assurance-test-cases.md]: sorry, i am unable to find the backend function. can to please give me the link to refer.

**Comment by khushishikhu** [comment]
@BadduCoder  can you please review my PR

**Comment by khushishikhu** [comment]
please someone review my pr

File: docs/quality-assurance-test-cases.md [hunk 6]
```diff
@@ -174,6 +174,33 @@ This document contains some examples of test cases for each feature implemented
...
+### List All Tasks
+
+**Service:** PUT /mentorship_relation/{request_id}/task/{task_id}
+
+| Test Case      | Outcome |
+| -------------- | -----------|
+| Logged in user requests list all task,or out of all task,for a mentorship relation those which involves the user (as a mentor or as a mentee) | Success |
+| Logged in user requests list all task,or out of all task, for a mentorship relation those which involves the user (neither as a mentor nor as a mentee) | Fail |
+| Logged in user requests to list all task for a invalid mentorship relation , which involves the user (as a mentor or as a mentee) | Fail |     
+| Logged in user requests list all task,or out of all task, those which doesn't involves the user (as a mentor or as a mentee) | Fail |
```
Comment (sumit-badsara) [inline docs/quality-assurance-test-cases.md]: Isn't this same as case 2?

File: docs/quality-assurance-test-cases.md [hunk 7]
```diff
@@ -174,6 +174,33 @@ This document contains some examples of test cases for each feature implemented
...
+| Not logged in user requests to List all task for any mentorship relation| Fail |
```
Comment (sumit-badsara) [inline docs/quality-assurance-test-cases.md]: for any **valid/invalid** mentorship relation

Just to make sure we also take care of invalid relation.

**Review by sumit-badsara — CHANGES_REQUESTED** [review]
Sorry I got a bit busy, couldn't review your PR. You are almost done. Just a tiny bit of change required.

**Comment by sumit-badsara** [comment]
Other than that @khushishikhu everything is almost done. 
Just check that 1 check is failing, take care of that and tag the issue creator (paritoshsinghrahar) or isabelcosta to approve and merge this.

**Comment by khushishikhu** [comment]
@paritoshsinghrahar please review my PR. and i don't why this 1 codecov check is not passing

File: docs/quality-assurance-test-cases.md [hunk 8]
```diff
@@ -174,6 +174,31 @@ This document contains some examples of test cases for each feature implemented
...
+| Create a task when a logged user is not involved in the relation                          | Fail    |
+| Create a task if relation state is different than accepted                                | Fail    |
+
+#### List All Tasks
+
+**Service:** GET /mentorship_relation/{request_id}/task
+
+| Test Case        | Outcome |
+| -------------- | ----------- |
+| Logged in user requests list all task,or out of all task,for a mentorship relation those which involves the user (as a mentor or as a mentee) | Success |
```
Comment (isabelcosta) [inline docs/quality-assurance-test-cases.md]: @khushishikhu I don't quite understand the sentence, what do you mean here?

Comment (khushishikhu) [reply docs/quality-assurance-test-cases.md]: @isabelcosta  i mean that the task will be successful when the logged in user which is either a mentor or a mentee requests to list it's all tasks for a mentorship relation . This is what i am getting that should be written over here, if there is still something inappropriate please let me know or guide me as well. Thank you

**Comment by paritoshsinghrahar** [comment]
> @paritoshsinghrahar please review my PR. and i don't why this 1 codecov check is not passing

I'll check and get back to you.

File: docs/quality-assurance-test-cases.md [hunk 9]
```diff
@@ -174,6 +174,31 @@ This document contains some examples of test cases for each feature implemented
...
+| Create a task for a relation, in the accepted state, between logged user and another user | Success |
+| Creating a task without a description (either empty or not in the request body at all)    | Fail    |
+| Create a task when a logged user is not involved in the relation                          | Fail    |
+| Create a task if relation state is different than accepted                                | Fail    |
+
+#### List All Tasks
+
+**Service:** GET /mentorship_relation/{request_id}/task
+
+| Test Case        | Outcome |
```
Comment (paritoshsinghrahar) [inline docs/quality-assurance-test-cases.md]: A nitpick. Space out the table in markdown to maintain uniformity with other tables.

Comment (khushishikhu) [reply docs/quality-assurance-test-cases.md]: @paritoshsinghrahar done with the changes. please have a look

Comment (paritoshsinghrahar) [reply docs/quality-assurance-test-cases.md]: Great.

**Comment by khushishikhu** [comment]
@isabelcosta  @paritoshsinghrahar  is this PR is okay or does it require still some work. please let me know

**Review by paritoshsinghrahar — DISMISSED** [review]

**Comment by rpattath** [comment]
Changing the status to `Needs Review` because there is only 1 review approval.

**Comment by devkapilbansal** [comment]
> Changing the status to `Needs Review` because there is only 1 review approval.

@rpattath ready to merge label should be removed too

**Comment by khushishikhu** [comment]
@devkapilbansal  can you please explain why this PR can't be merged, is there anything wrong in this PR? Because it is already approved by OSH mentor. Please elaborate me what else changes needs to be done in this PR to get merged

**Comment by devkapilbansal** [comment]
@khushishikhu every PR needs atleast two approvals before moving towards testing or Rrady to Merge. As this PR is reviewed by one member only, this PR needs review.
Also, as reviews are required on this pr, this PR can't be merged at this moment

File: docs/quality-assurance-test-cases.md [hunk 10]
```diff
@@ -174,6 +174,31 @@ This document contains some examples of test cases for each feature implemented
...
+| Create a task if relation state is different than accepted                                | Fail    |
+
+#### List All Tasks
+
+**Service:** GET /mentorship_relation/{request_id}/task
+
+| Test Case                                                                                                                                               | Outcome |
+| ------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- |
+| Logged in user requests list all task,or out of all task,for a mentorship relation those which involves the user (as a mentor or as a mentee)           | Success |
+| Logged in user requests list all task,or out of all task, for a mentorship relation those which involves the user (neither as a mentor nor as a mentee) | Fail |
```
Comment (devkapilbansal) [inline docs/quality-assurance-test-cases.md]:
```suggestion
| Logged in user requests list all task,or out of all task, for a mentorship relation which does not involves the user (neither as a mentor nor as a mentee) | Fail |
```

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
@khushishikhu please rebase your branch with current branch

**Comment by epicadk** [comment]
@khushishikhu please remember to use a different branch for each issue next time. : )

File: docs/quality-assurance-test-cases.md (hunk 10 above)
Comment (devkapilbansal) [reply docs/quality-assurance-test-cases.md]: @khushishikhu a minor change needed here

**Comment by khushishikhu** [comment]
> @khushishikhu please remember to use a different branch for each issue next time. : ) 

@epicadk  yes sure. Actually by mistake I have added changes without concern about the branch 😅

File: docs/quality-assurance-test-cases.md (hunk 10 above)
Comment (vj-codes) [reply docs/quality-assurance-test-cases.md]: @khushishikhu please update as suggested

File: docs/quality-assurance-test-cases.md [hunk 11]
```diff
@@ -174,6 +174,31 @@ This document contains some examples of test cases for each feature implemented
...
+
+**Service:** POST /mentorship_relation/{relation_id}/task
+
+|  Test Case                                                                                | Outcome |
+| ----------------------------------------------------------------------------------------- |-------- |
+| Create a task for a relation, in the accepted state, between logged user and another user | Success |
+| Creating a task without a description (either empty or not in the request body at all)    | Fail    |
+| Create a task when a logged user is not involved in the relation                          | Fail    |
+| Create a task if relation state is different than accepted                                | Fail    |
+
```
Comment (isabelcosta) [inline docs/quality-assurance-test-cases.md:189]: this is already present on the `develop` branch of the project. Is your fork up to date with this repository @khushishikhu ?

Comment (isabelcosta) [reply docs/quality-assurance-test-cases.md:189]: I see this is duplicated in this file, the same table is present on line 231, could you please remove the duplicate from below this line? So you only have this page here?

**Comment by vj-codes** [comment]
@khushishikhu any updates on this?

File: docs/quality-assurance-test-cases.md (hunk 11 above)
Comment (khushishikhu) [reply docs/quality-assurance-test-cases.md:189]: yes sure I will remove the duplicate table from that line.Thank you!

**Comment by khushishikhu** [comment]
yes @vj-codes I have done the changes as suggested before

File: docs/quality-assurance-test-cases.md [hunk 12]
```diff
@@ -174,6 +174,31 @@ This document contains some examples of test cases for each feature implemented
...
+#### List All Tasks
+
+**Service:** GET /mentorship_relation/{request_id}/task
+
+| Test Case                                                                                                                                               | Outcome |
+| ------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- |
+| Logged in user requests list all task,or out of all task,for a mentorship relation those which involves the user (as a mentor or as a mentee)           | Success |
+| Logged in user requests list all task,or out of all task, for a mentorship relation which does not involves the user (neither as a mentor nor as a mentee) | Fail |
+| Logged in user requests to list all task for a invalid mentorship relation , which involves the user (as a mentor or as a mentee)                       | Fail |
```
Comment (devkapilbansal) [inline docs/quality-assurance-test-cases.md:198]: Which  case is this? If relation is invalid, then how the user can be involved with it?

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
@khushishikhu please change all instances of `logged in user` with `authenticated user`.
Also, I have a minor dobut here :point_down:

**Comment by devkapilbansal** [comment]
@khushishikhu are you still working on this??

**Comment by vj-codes** [comment]
@khushishikhu closing the PR due to inactivity, thank you for your contribution :)
Let us know if you would like to continue again

**Pull request closed without merging**